webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js

```diff
@@ -0,0 +1,135 @@
+  const response = useSelector(state => selectCVHistories(state, cvId));
+  const status = useSelector(state => selectCVHistoriesStatus(state, cvId));
+  const error = useSelector(state => selectCVHistoriesError(state, cvId));
+  const loading = status === STATUS.PENDING;
+
+  const [rows, setRows] = useState([]);
+  const [metadata, setMetadata] = useState({});
+  const [searchQuery, updateSearchQuery] = useState('');
+
+  const taskTypes = {
+    publish: 'Actions::Katello::ContentView::Publish',
+    promotion: 'Actions::Katello::ContentView::Promote',
+    removal: 'Actions::Katello::ContentView::Remove',
+    incrementalUpdate: 'Actions::Katello::ContentView::IncrementalUpdates',
+    export: 'Actions::Katello::ContentViewVersion::Export',
+  };
+
+  const columnHeaders = [
+    __('Date'),
+    __('Version'),
+    __('Status'),
+    __('Action'),
+    __('Description'),
+    __('User'),
+  ];
+
+  const actionText = (history) => {
+    const {
+      action,
+      task,
+      environment,
+    } = history;
+
+    const taskType = task ? task.label : taskTypes[action];
+
+    if (taskType === taskTypes.removal) {
+      return <React.Fragment> Deleted from <Label key="1" color="blue" href={`/lifecycle_environments/${environment.id}`}>{`${environment.name}`}</Label>{}</React.Fragment>;
+    } else if (taskType === taskTypes.promotion) {
+      return <React.Fragment> Promoted to <Label key="2" color="blue" href={`/lifecycle_environments/${environment.id}`}>{`${environment.name}`}</Label>{}</React.Fragment>;
+    } else if (taskType === taskTypes.publish) {
+      return ('Published new version');
+    } else if (taskType === taskTypes.export) {
+      return ('Exported content view');
+    } else if (taskType === taskTypes.incrementalUpdate) {
+      return ('Incremental update');
+    }
+    return '';
+  };
+
+  const buildRows = (results) => {
+    const newRows = [];
+    results.forEach((history) => {
+      const {
+        version,
+        version_id: versionId,
+        created_at: createdAt,
+        status: taskStatus,
+        description,
+        user,
+      } = history;
+
+      const actionMessage = actionText(history);
+
+      const cells = [
+        { title: <LongDateTime date={createdAt} showRelativeTimeTooltip /> },
+        { title: <a href={urlBuilder(`content_views/${cvId}/versions/${versionId}`, '')}>{`Version ${version}`}</a> },
+        taskStatus,
+        actionMessage,
+        { title: <TableText wrapModifier="truncate">{description}</TableText> },
+        user,
+      ];
+
+      newRows.push({ cells });
+    });
+    return newRows;
+  };
+
+  useEffect(() => {
+    const { results, ...meta } = response;
+    setMetadata(meta);
+    if (!loading && results) {
+      const newRows = buildRows(results);
+      setRows(newRows);
+    }
+  }, [JSON.stringify(response)]);
+
+  const emptyContentTitle = __("You currently don't have any history for this content view.");
```

Comment:
Noticed this in our meet today: Should make sure the empty state isn't displayed while it's loading.  (should have a spinner instead)

Reply:
https://projects.theforeman.org/issues/32229